lib/parser/source/buffer.rb

```diff
@@ -303,27 +296,49 @@ def source_range
       # @return [Integer]
       #
       def last_line
-        line_begins.size + @first_line - 1
+        line_begins.size + @first_line - 2
+      end
+
+      # :nodoc:
+      def freeze
```

Comment:
I don't know if I agree with this. I mean, I understand the rationale behind it, but I'm just afraid that MRI can implicitly call this method in different places without going through dynamic dispatch, i.e. it can call `rb_freeze` (of how's it called) to set a bitflag directly.

It's better than what's currently in master, but it's a source of potential confusion.

If these ivars are set in most cases we could compute them in the constructor. If not - I guess I'm ok with overriding `.freeze`.

Also, side question: why do you want buffer to be frozen?

Reply:
> I'm just afraid that MRI can implicitly call this method in different places without going through dynamic dispatch

AFAIK, there is no such circumstance currently; Ruby never sets the freeze flag directly. I've been advocating so that there isn't, so far so good and `Ractor.make_shareable` calls user-defined `freeze` methods (and then proceeds to deep freeze by calling `#freeze` on its unfrozen components).

> why do you want buffer to be frozen?

Currently, there is no good reason to. 
I understand that it's typical in Ruby to not "care" about `freeze` (or `initialize_copy` for mutable), but it is part of the public API, and currently if someone calls it, it may crash.

An alternative implementation would be to store the caches inside a struct. This way we could `freeze` the buffer when it is completed with a source, and not need to prebuild the caches (unless the struct itself gets frozen, a sign that we are being deep-frozen). This would reflect the fact that, once the source is given, the buffer is immutable (`frozen? # => true`).

Reply:
> if someone calls it, it may crash

I'd not expect anyone to freeze objects that they don't own, but I agree with your point.